Interactivity API: Refactor context proxies

DAreRodz opened the pull request:
## What?

Minor refactoring for the context proxies. This PR moves the related code to the `/proxies` folder inside the Interactivity API package and removes unnecessary code.

Follow up of https://github.com/WordPress/gutenberg/pull/62734.

## Why?

Code quality and maintainability.

## How?

This PR keeps the previous logic; this means context objects still have two proxies. It's a bit tricky to merge the state proxies logic with the context proxies one, so we could address that in a future PR.

## Testing Instructions

Current tests should pass as expected.

DAreRodz committed:
Move `proxifyContext` to proxies folder

DAreRodz committed:
Move context handlers outside

DAreRodz committed:
Remove unused code

DAreRodz committed:
Update comment

DAreRodz committed:
Add TODO comment

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+// TODO: Use the proxy registry to avoid multiple proxies on the same object.
```
What is this?

DAreRodz replied:
It's just a reminder of what I mentioned above in the description:

> It's a bit tricky to merge the state proxies logic with the context proxies one, so we could address that in a future PR.

luisherranz replied:
Ok, thanks! But, to be honest, I wouldn't note that in the code…

DAreRodz replied:
I've moved it above `contextHandlers`.

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+	get: ( target: object, k: string ) => {
+		const fallback = contextObjectToFallback.get( target );
+		// Always subscribe to prop changes in the current context.
+		const currentProp = target[ k ];
```
Why not use `Reflect.get` here?

DAreRodz replied:
It would be equivalent. 😄

In this case, we don't want to pass the context proxy as the receiver, so we would have to call `Reflect.get` like this:

```js
const currentProp = Reflect.get( target, k );
```

I can change it if you think it is more _correct_, though.

luisherranz replied:
> we don't want to pass the context proxy as the receiver

How so?

DAreRodz replied:
Because it changes the receiver used inside the state proxy handlers to be the context proxy one.

You can test it with this example. It makes an object to return the receiver for any property.

```js
const obj = {}
const proxyA = new Proxy(
    obj,
    {
        get( _t, _k, receiver ) {
            return receiver; // We expect `receiver` to be `proxyA` here.
        }
    }
);

proxyA.prop === proxyA // true

const proxyB = new Proxy(
    proxyA, // It wraps proxyA.
    {
        get( t, k, receiver ) {
            // `t` is proxyA`, but we pass `proxyB` as the receiver
            return Reflect.get( t, k, receiver );
        }
    }
);


proxyB.prop === proxyA // false (should be true)
proxyB.prop === proxyB // true  (should be false)
```

DAreRodz replied:
We're using proxy references to point to other related data, like reactive props, so we need to maintain the correct proxy reference (the state proxy one). 🙂

luisherranz replied:
That's interesting; thanks, David, for the explanation.

luisherranz replied:
Resolving this then.

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+const descriptor = Reflect.getOwnPropertyDescriptor;
+
+const contextHandlers: ProxyHandler< object > = {
+	get: ( target: object, k: string ) => {
```
It might be better to use `key` instead of `k` here to follow the same nomenclature as in the rest of the handlers.

DAreRodz replied:
Right, I'll do that.

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+		// Always subscribe to prop changes in the current context.
+		const currentProp = target[ k ];
+
+		// Return the inherited prop when missing in target.
+		if ( ! ( k in target ) && k in fallback ) {
+			return fallback[ k ];
+		}
+
+		/*
+		 * For other cases, return the value from target, also
+		 * subscribing to changes in the parent context when the current
+		 * prop is not defined.
+		 */
+		return k in target ? currentProp : fallback[ k ];
```
What is the difference between these two? Why do we need the first one?

DAreRodz replied:
The one above is the case when a property is not defined in `target` but in `fallback`. In this case, it should subscribe to the signal from `fallback`.

The one below covers the following two cases:

1. The property is defined in `target`.
2. ~The property is neither defined in `target` nor in `fallback`~

For both, the signal used should come from `target`.

EDIT: it should return the prop from `fallback` in the second case. My bad. 😅

luisherranz replied:
But if you remove the `if` above, for that condition, the return below also returns `fallback[ k ]`, doesn't it? 

> For both, the signal used should come from target.

But you are returning the fallback when the property is not in `target`: `return k in target ? currentProp : fallback[ k ];`

Or what am I missing? 😄

DAreRodz replied:
Oh, that's true. The implementation could be wrong. I'll revisit this part.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/64713/commits/cc3d640c421d378851ebcad942841d8f2ab9b18d.

I was confused about the logic that should run there. It should return `fallback[ k ]` when the property is neither defined in target nor in fallback.

luisherranz replied:
Cool, thanks.

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+	set: ( target, k, value ) => {
+		const fallback = contextObjectToFallback.get( target );
+		const obj = k in target || ! ( k in fallback ) ? target : fallback;
+		obj[ k ] = value;
```
`Reflect.set` here as well?

DAreRodz replied:
Same answer as in https://github.com/WordPress/gutenberg/pull/64713#discussion_r1728853160. 😄

luisherranz commented on packages/interactivity/src/proxies/context.ts:
```diff
@@ -0,0 +1,69 @@
+	if ( ! contextObjectToProxy.has( current ) ) {
+		const proxy = new Proxy( current, contextHandlers );
+		contextObjectToProxy.set( current, proxy );
+		contextProxyToObject.set( proxy, current );
```
Is `contextProxyToObject` still necessary?

DAreRodz replied:
It's not. Nice catch. I'll remove it.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/64713/commits/cc3d640c421d378851ebcad942841d8f2ab9b18d.

luisherranz reviewed (commented):
It looks good so far. Thanks for refactoring this, @DAreRodz!

I have left some minor comments and questions here.

DAreRodz committed:
Remove unused code and rename `k` arg to `key`

luisherranz commented:
Thanks, David. Let me know once it's ready for review.

DAreRodz committed:
Replace `updateContext` with `deepMerge`

DAreRodz committed:
Throw when calling `proxifyContext` on a context proxy

DAreRodz committed:
Add unit tests

DAreRodz committed:
Update changelog

DAreRodz commented:
The error that you see in the `wp-each` tests ([here](https://github.com/WordPress/gutenberg/actions/runs/10762188484/job/29842221199?pr=64713#step:7:28)) should be gone once the bug fix in https://github.com/WordPress/gutenberg/pull/65135 is merged. ☝️

michalczaplinski commented on packages/interactivity/src/directives.tsx:
```diff
@@ -24,131 +20,7 @@ import {
-				// Always subscribe to prop changes in the current context.
-				const currentProp = target[ k ];
-
-				// Return the inherited prop when missing in target.
-				if ( ! ( k in target ) && k in fallback ) {
-					return fallback[ k ];
-				}
-
-				// Proxify plain objects that were not directly assigned.
-				if (
-					k in target &&
-					! contextAssignedObjects.get( target )?.has( k ) &&
-					isPlainObject( currentProp )
-				) {
-					return proxifyContext( currentProp );
-				}
-
-				// Return the stored proxy for `currentProp` when it exists.
-				if ( contextObjectToProxy.has( currentProp ) ) {
-					return contextObjectToProxy.get( currentProp );
-				}
```
Why do we not need this logic anymore?

DAreRodz replied:
See https://github.com/WordPress/gutenberg/pull/64713#discussion_r1750580817.

michalczaplinski commented on packages/interactivity/src/directives.tsx:
```diff
@@ -286,9 +158,10 @@ export default () => {
 						currentValue.current,
-						deepClone( value ) as object
+						deepClone( value ) as object,
+						false
```
It's not clear to me that `deepMerge` is equivalent to the `updateContext()` function. Should it be equivalent or are we consciously changing the logic here? If it's the second one, then why do we do that?

DAreRodz replied:
The idea here is to use `deepMerge()` with the `override` option set to `false`. This algorithm follows the same logic as `updateContext` but does not manipulate the reactive object directly and only updates the signals created beforehand.

That way, we avoid `wp-context` subscribing unnecessarily to the modified context and prevent wasteful signal instantiation. 🙂

michalczaplinski replied:
ok, I see. Upon further inspection the logic does seem **almost** the same. The only 2 differences I see is:

1. We do not use `peek()` anymore. This I assume is because we the context is not wrapped in proxies anymore like you mentioned [here](https://github.com/WordPress/gutenberg/pull/64713#discussion_r1750588651).

2. I see that we're now assigning empty object here:
https://github.com/WordPress/gutenberg/blob/ce56700594dd7cb8b1794488e2c010706d310b81/packages/interactivity/src/proxies/state.ts#L298-L301

   but [previously](https://github.com/WordPress/gutenberg/pull/64713/files#diff-118c339960e17dbd5b9434e6a2eec6851c8b8f68b6ace6faf4e57603850eb169L147-L149) we used to re-assign the property:

   ```js
   } else if ( ! ( k in target ) ) {
     target[ k ] = source[ k ];
   }
   ```

DAreRodz replied:
> 1. We do not use `peek()` anymore. This I assume is because we the context is not wrapped in proxies anymore like you mentioned [here](https://github.com/WordPress/gutenberg/pull/64713#discussion_r1750588651).

Exactly. 🙂 



> 2. I see that we're now assigning empty object here [...] but [previously](https://github.com/WordPress/gutenberg/pull/64713/files#diff-118c339960e17dbd5b9434e6a2eec6851c8b8f68b6ace6faf4e57603850eb169L147-L149) we used to re-assign the property:

That was wrong. We don't want to copy a plain object's reference from `source` when calling `deepMerge`.

michalczaplinski commented on packages/interactivity/src/directives.tsx:
```diff
@@ -24,131 +20,7 @@ import {
-				 * to the original object and don't inherit unexpected
-				 * properties.
-				 */
-				if ( value && typeof value === 'object' ) {
-					if ( ! contextAssignedObjects.has( obj ) ) {
-						contextAssignedObjects.set( obj, new Set() );
-					}
-					contextAssignedObjects.get( obj ).add( k );
-				}
-
-				/*
-				 * When the value is a proxy, it's because it comes from the
-				 * context, so the inner value is assigned instead.
-				 */
-				if ( contextProxyToObject.has( value ) ) {
-					const innerValue = contextProxyToObject.get( value );
-					obj[ k ] = innerValue;
-				} else {
-					obj[ k ] = value;
-				}
```
Likewise, why do we not need this logic anymore?

DAreRodz replied:
That's because objects at that level in the context are no longer wrapped with context proxies, so they keep their original references. This means we can eliminate the logic that checks whether an object should be _proxified_ or not.

michalczaplinski replied:
Why are they not wrapped with context proxies anymore at that level? Could you explain :)

DAreRodz replied:
Properties from parent contexts were inherited from nested objects, making dealing with the context unnecessarily complex, especially when trying to avoid inheriting unknown properties. Since https://github.com/WordPress/gutenberg/pull/62734, only the properties defined at the root level of the parent context are inherited.

michalczaplinski replied:
👍

michalczaplinski committed:
Added new tests
to validate behavior with proxified state. Also included more descriptive error messages for re-proxification attempts.

michalczaplinski reviewed (approved):
Nice work @DAreRodz ! 

This looks good to me! 👍 I hope it's okay I pushed a tiny improvement to the tests.
